### PR TITLE
feat(Inference): Model gives back stacked tensor of scores each row a user, each column an article.

### PR DESCRIPTION
**Pull request by szemyd**
- Important! We are padding the tensors to be the same shape with a very large negative number! This ensures the padding won't be chosen during topK.

### CODE REVIEW

File: data/data_loader.py [hunk 1]
```diff
@@ -39,8 +39,8 @@ def create_dataloaders(
     )
 
     train_loader = DataLoader(train_dataset, batch_size=config.batch_size, shuffle=True)
-    val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)
-    test_loader = DataLoader(test_dataset, batch_size=1, shuffle=False)
+    val_loader = DataLoader(val_dataset, batch_size=config.batch_size, shuffle=True)
```
Comment (szemyd) [inline data/data_loader.py:42]: I have reintroduced shuffleing to the validation and test set.
We have now batching here aswell.

File: model/encoder_decoder.py [hunk 2]
```diff
@@ -126,7 +127,23 @@ def forward(
...
+            user_id = user_index.item()
+            score = out[i].unsqueeze(0)
+            if user_id >= len(out_per_user):
+                out_per_user.append(score)
+            else:
+                out_per_user[user_id] = torch.concat([out_per_user[user_id], score])
+
+        max_output_length = max([element.shape[0] for element in out_per_user])
+        
+        padded_tensors = [F.pad(output, (0,max_output_length-output.shape[0]), "constant", -(1 << 50)) for output in out_per_user]
```
Comment (szemyd) [inline model/encoder_decoder.py:145]: I hope this is not too inefficient.

File: single_epoch.py [hunk 3]
```diff
@@ -69,15 +69,15 @@ def epoch_with_dataloader(
...
+        if i == 50:
             break
         val_loop.set_description(f"Val, epoch: {epoch_id}")
         val_recall, val_precision = test(data.to(device), model, [])
         val_loop.set_postfix_str(f"Recall Val: {val_recall:.4f}")
 
     test_loop = tqdm(iter(test_loader))
     for i, data in enumerate(test_loop):
-        if i == 5:
+        if i == 50:
```
Comment (szemyd) [inline single_epoch.py:80]: We will have to delete these in the final run.

File: utils/metrics_encoder_decoder.py [hunk 4]
```diff
@@ -137,7 +137,11 @@ def get_metrics_universal(
     """
 
     # Ratings expects a tensor of ratings for multiple users
-    ratings = model_output.unsqueeze(0)
+
+    if len(model_output.shape) < 2:
+        ratings = model_output.unsqueeze(0)
+    else:
+        ratings = model_output
```
Comment (szemyd) [inline utils/metrics_encoder_decoder.py:144]: Maybe this is superflous now? This was needed when batch was 1, but we have changed the output anyway so it might be alright?

File: single_epoch.py (hunk 3 above)
Comment (itchingpixels) [reply single_epoch.py:80]: should we make this a config parameter? "evaluate_break_at"?

**Review by itchingpixels — APPROVED** [review]
👍
